### PR TITLE
Create Element and Nodal Reporters

### LINKED ISSUE #24678: Extend Reporter System
## Reason

Extend reporter systems to have ElementReporters and NodalReporters. This makes aggregating data easier when needing to loop over nodes or elements.
## Design

Derive ElementReporters and NodalReporters from GeneralReporters and Element/NodalUserObjects.
## Impact

Adds new capabilities without changing any existing APIs.

### PR DESCRIPTION
**Pull request by maxnezdyur**
closes #24678 
## Reason

Will allow for easy reporter management when the reporter depends on a set of elements or nodes. Will be specifically used in the optimization module.
## Design

Derive off of Element/Nodal UserObjects and Reporter

## Impact

Increases Reporter usability when needing a reporter that only works on a block, boundary, etc.

### CODE REVIEW

**Comment by maxnezdyur** [comment]
I don't think reporters will work with threading? @zachmprince

File: test/tests/reporters/ele_nodal_reporters/element_reporter/tests
Comment (zachmprince) [inline test/tests/reporters/ele_nodal_reporters/element_reporter/tests:1]: Let's remove the `ele_nodal_reporters` folder and just have `element_reporter` and `nodal_reporter`.

**Review by zachmprince — CHANGES_REQUESTED** [review]

File: framework/src/reporters/CoupledVarStatsElementReporter.C [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "CoupledVarStatsElementReporter.h"
+
+registerMooseObject("MooseApp", CoupledVarStatsElementReporter);
```
[the remark this answers is not in the document]
Comment (maxnezdyur) [reply framework/src/reporters/CoupledVarStatsElementReporter.C:12]: No apologies, don't like the name either. So `ElementVariableStatistics` and the base class `ElementStatistics`?

**Comment by maxnezdyur** [comment]
I see the problem. In 
```
void
FEProblemBase::addUserObject(const std::string & user_object_name,
                             const std::string & name,
                             InputParameters & parameters)
{
  // Add the _subproblem and _sys parameters depending on use_displaced_mesh
  addObjectParamsHelper(parameters, name);

  UserObject * primary = nullptr;
  for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
  {
    // Create the UserObject
    std::shared_ptr<UserObject> user_object =
        _factory.create<UserObject>(user_object_name, name, parameters, tid);
    if (tid == 0)
      primary = user_object.get();
    else
      user_object->setPrimaryThreadCopy(primary);

    // TODO: delete this line after apps have been updated to not call getUserObjects
    _all_user_objects.addObject(user_object, tid);

    theWarehouse().add(user_object);

    // Attempt to create all the possible UserObject types
    auto euo = std::dynamic_pointer_cast<ElementUserObject>(user_object);
    auto suo = std::dynamic_pointer_cast<SideUserObject>(user_object);
    auto isuo = std::dynamic_pointer_cast<InternalSideUserObject>(user_object);
    auto iuob = std::dynamic_pointer_cast<InterfaceUserObjectBase>(user_object);
    auto nuo = std::dynamic_pointer_cast<NodalUserObject>(user_object);
    auto duo = std::dynamic_pointer_cast<DomainUserObject>(user_object);
    auto guo = std::dynamic_pointer_cast<GeneralUserObject>(user_object);
    auto tguo = std::dynamic_pointer_cast<ThreadedGeneralUserObject>(user_object);
    auto muo = std::dynamic_pointer_cast<MortarUserObject>(user_object);

    // Account for displaced mesh use
    if (_displaced_problem && parameters.get<bool>("use_displaced_mesh"))
    {
      if (euo || nuo || duo)
        _reinit_displaced_elem = true;
      else if (suo)
        // shouldn't we add isuo
        _reinit_displaced_face = true;
      else if (iuob)
        _reinit_displaced_neighbor = true;
    }

    if ((guo && !tguo) || muo)
      break;
  }
} 
```
 we break when for general user objects which is why general reporter doesn't have this problem. Not sure about the best fix but I could add a member to `UserObject` that defaults to true, but can be made to be false for reporters and use that to break out.

**Comment by maxnezdyur** [comment]
From my initial tests, I think I see two ways to do this.
1) Having only the "root" thread declare the reporter but still added the UO on the other threads and then in `threadJoin` making sure we save the data correctly.
2) Somehow having these reporterUO not go into the threadedUO loops and just have one reporterUO that loops over all the elements/nodes. 

The problem with my initial solution (break out of the for loop for the ele/nodalReporters) was that the UO was only added for one "thread" and then the reporter would only get 1/n data, even with threadJoin overrided to correctly add to the reporter. Will try to look at the least invasive way to do this. 
Is there some way to force a Element/NodalUO to not be threaded?

File: framework/src/outputs/JSONOutput.C
Comment (maxnezdyur) [inline framework/src/outputs/JSONOutput.C:1]: We need to create multiple objects for the threading capabilities but only the first object contains the correct data. Not sure how to error check  multiple reporters with same name but not of the threaded variety.

File: framework/include/reporters/Reporter.h
Comment (maxnezdyur) [inline framework/include/reporters/Reporter.h:1]: Use the UnusedValue to provide provide a reference for the threaded objects that will not get declared. Not sure if there is a better way to do this. This is very similar to how vectorpostprocessors work in declareValue.

File: framework/include/reporters/ElementVariableStatistics.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  ElementVariableStatistics(const InputParameters & parameters);
+
+private:
+  /// The coupled variable used.
+  const VariableValue & _v;
+  virtual Real computeValue();
```
Comment (zachmprince) [inline framework/include/reporters/ElementVariableStatistics.h:24]:
```suggestion
  virtual Real computeValue() override;

    /// The coupled variable used.
  const VariableValue & _v;
```

File: framework/include/reporters/NodalVariableStatistics.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+{
+public:
+  static InputParameters validParams();
+
+  NodalVariableStatistics(const InputParameters & parameters);
+
+private:
+  /// The coupled variable used.
+  const VariableValue & _v;
+  virtual Real computeValue();
```
Comment (zachmprince) [inline framework/include/reporters/NodalVariableStatistics.h:24]:
```suggestion
  virtual Real computeValue() override;

    /// The coupled variable used.
  const VariableValue & _v;
```

File: framework/src/reporters/ElementStatistics.C [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _max(declareValueByName<Real>(_base_name + "max")),
+    _min(declareValueByName<Real>(_base_name + "min")),
+    _average(declareValueByName<Real>(_base_name + "average")),
+    _integral(declareValueByName<Real>(_base_name + "integral")),
+    _number_elements(declareValueByName<int>(_base_name + "number_elements"))
+{
+}
+void
+ElementStatistics::initialize()
```
Comment (zachmprince) [inline framework/src/reporters/ElementStatistics.C:31]: I'm surprised formatting didn't pick this up
```suggestion
{
}

void
ElementStatistics::initialize()
```

File: framework/src/reporters/ElementVariableStatistics.C [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+                             "be transfered to other apps.");
+  return params;
+}
+
+ElementVariableStatistics::ElementVariableStatistics(const InputParameters & parameters)
+  : ElementStatistics(parameters), _v(coupledValue("coupled_var"))
+{
+}
+Real
+ElementVariableStatistics::computeValue()
```
Comment (zachmprince) [inline framework/src/reporters/ElementVariableStatistics.C:31]:
```suggestion
{
}

Real
ElementVariableStatistics::computeValue()
```

File: framework/src/reporters/NodalStatistics.C [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  : NodalReporter(parameters),
+    _base_name(isParamValid("base_name") ? getParam<std::string>("base_name") + "_" : ""),
+    _max(declareValueByName<Real>(_base_name + "max")),
+    _min(declareValueByName<Real>(_base_name + "min")),
+    _average(declareValueByName<Real>(_base_name + "average")),
+    _number_nodes(declareValueByName<int>(_base_name + "number_nodes"))
+{
+}
+void
+NodalStatistics::initialize()
```
Comment (zachmprince) [inline framework/src/reporters/NodalStatistics.C:30]:
```suggestion
{
}

void
NodalStatistics::initialize()
```

File: framework/src/reporters/NodalVariableStatistics.C [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                             "be transfered to other apps.");
+  return params;
+}
+
+NodalVariableStatistics::NodalVariableStatistics(const InputParameters & parameters)
+  : NodalStatistics(parameters), _v(coupledValue("coupled_var"))
+{
+}
+Real
+NodalVariableStatistics::computeValue()
```
Comment (zachmprince) [inline framework/src/reporters/NodalVariableStatistics.C:31]:
```suggestion
}

Real
NodalVariableStatistics::computeValue()
```

File: test/tests/reporters/element_reporter/elem_stats.i [hunk 8]
```diff
@@ -0,0 +1,57 @@
+[Mesh]
+  type = GeneratedMesh
+  dim = 2
+  xmax = 2
+  ymax = 2
+  nx = 10
+  ny = 10
+[]
```
Comment (zachmprince) [inline test/tests/reporters/element_reporter/elem_stats.i:8]:
```suggestion
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = 2
    ymax = 2
    nx = 10
    ny = 10
  []
[]
```

File: test/tests/reporters/nodal_reporter/nodal_stats.i [hunk 9]
```diff
@@ -0,0 +1,57 @@
+[Mesh]
+  type = GeneratedMesh
+  dim = 2
+  xmax = 2
+  ymax = 2
+  nx = 10
+  ny = 10
+[]
```
Comment (zachmprince) [inline test/tests/reporters/nodal_reporter/nodal_stats.i:8]:
```suggestion
[Mesh]
  [gmg]
    type = GeneratedMeshGenerator
    dim = 2
    xmax = 2
    ymax = 2
    nx = 10
    ny = 10
  []
[]
```

**Review by zachmprince — CHANGES_REQUESTED** [review]
Great work!! I'm impressed you figured out the threading issue. Just a few minor requests, then I'll merge.

**Review by zachmprince — APPROVED** [review]

**Pull request merged**